packages/vdom-extension/src/index.tsx

```diff
@@ -43,27 +45,71 @@ export class RenderedVDOM extends Widget implements IRenderMime.IRenderer {
     this.addClass('jp-RenderedHTML');
     this.addClass('jp-RenderedHTMLCommon');
     this._mimeType = options.mimeType;
+    // Get current kernel session (hack for mimerender extension)
+    this._session = (options.resolver as any)._session;
   }
 
   /**
-   * Render VDOM into this widget's node.
+   * Dispose of the widget.
    */
-  renderModel(model: IRenderMime.IMimeModel): Promise<void> {
-    const data = model.data[this._mimeType] as any;
-    return new Promise<void>(resolve => {
-      ReactDOM.render(<VDOM data={data} />, this.node, resolve);
-    });
+  dispose(): void {
+    // Dispose of comm disposables
+    for (let targetName in this._comms) {
+      this._comms[targetName].dispose();
+    }
+    super.dispose();
   }
 
   /**
    * Called before the widget is detached from the DOM.
    */
   protected onBeforeDetach(msg: Message): void {
-    // Unmount the component so it can tear down.
+    // Dispose of React component(s).
     ReactDOM.unmountComponentAtNode(this.node);
   }
 
+  /**
+   * Render VDOM into this widget's node.
+   */
+  renderModel(model: IRenderMime.IMimeModel): Promise<void> {
+    return new Promise((resolve, reject) => {
+      const data = model.data[this._mimeType] as any;
+      // const metadata = model.metadata[this._mimeType] as any || {};
+      ReactDOM.render(
+        <VDOM data={data} onVDOMEvent={this.handleVDOMEvent} />,
+        this.node,
+        () => {
+          resolve();
+        }
+      );
+    });
+  }
+
+  /**
+   * Handle events for VDOM element.
+   */
+  handleVDOMEvent = (targetName: string, event: SerializedEvent<any>): void => {
+    // When a VDOM element's event handler is called, send a serialized
+    // representation of the event to the registered comm channel for the
+    // kernel to handle
+    if (this._timer) {
+      window.clearTimeout(this._timer);
+    }
+    this._timer = window.setTimeout(() => {
+      if (!this._comms[targetName]) {
+        this._comms[targetName] = this._session.kernel.connectToComm(
+          targetName
+        );
+        this._comms[targetName].open();
+      }
+      this._comms[targetName].send(JSON.stringify(event));
+    }, 16);
```

Comment:
Why 16 ms?

Reply:
1000 ms / 60 fps = 16 ms per frame

This could certainly be increased, I don't think anything under 60ms is observable to humans. We could also try using [requestAnimationFrame](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame).

Reply:
Am I right in noting that this logic will drop events that happen within 16ms of a succeeding event? In general that seems fine to me, just trying to understand the UX implications here.

Reply:
Yes, this is to debounce the sending of comm messages. You can see the effect of too many comm messages when you try the _scroll_ example in the demo notebook.

Reply:
This has proven to be problematic for a use case where you are handling both `onMouseMove` and `onMouseUp` or `onMouseDown` events because mousemove events happen so frequently that mouseup or mousedown events can be dropped. I'm going to bump this back down to 16ms perhaps we should drop it altogether?